alacritty_terminal/src/term/cell.rs

```diff
@@ -94,25 +96,50 @@ impl Cell {
 
     /// Write a new zerowidth character to this cell.
     #[inline]
-    pub fn push_zerowidth(&mut self, c: char) {
-        self.extra.get_or_insert_with(Default::default).zerowidth.push(c);
-    }
+    pub(crate) fn push_zerowidth(&mut self, character: char) {
+        self.init_extra();
 
-    /// Free all dynamically allocated cell storage.
-    #[inline]
-    pub fn drop_extra(&mut self) {
-        if self.extra.is_some() {
-            self.extra = None;
-        }
+        let extra = Arc::make_mut(self.extra.as_mut().unwrap());
+        extra.zerowidth.push(character);
     }
 
     /// Remove all wide char data from a cell.
     #[inline(never)]
-    pub fn clear_wide(&mut self) {
+    pub(crate) fn clear_wide(&mut self) {
         self.flags.remove(Flags::WIDE_CHAR);
-        self.drop_extra();
+        self.extra = None;
         self.c = ' ';
     }
+
+    /// Set underline color on the cell.
+    pub(crate) fn set_underline_color(&mut self, color: Option<Color>) {
+        // If we reset color and we don't have zerowidth we should drop extra storage.
+        if color.is_none()
+            && self.extra.as_ref().map(|extra| !extra.zerowidth.is_empty()) != Some(false)
+        {
+            self.extra = None;
+            return;
+        }
+
+        self.init_extra();
+
+        let extra = Arc::make_mut(self.extra.as_mut().unwrap());
```

Comment:
@chrisduerr you can't mutate value if there're references to it. Make mute checks for strong references and does a copy if there're more than one. There's no need to copy anything if the reference is unique.

It's rare for underline, but it's common for zerowidth or when you add underline to zerowidth.

Reply:
Yeah that's fair I just wonder which one is more efficient? Because zerowidth should be fairly rare. I feel like both are just slightly different trade-offs with their own advantages?

Reply:
idk, I like `make_mut`, since it's safer and more intuitive. Checking 2 counters isn't that costly.